ichnaea/data/tests/test_station.py

```diff
@@ -137,10 +151,150 @@ def mock_area(self, updated_areas, key, num=num, session2=session2):
+                "Deadlock found when trying to get lock; try restarting transaction",
+            ),
+            (
+                # This is an OperationalError in PyMySQL 0.9.0 and later
+                OperationalError,
+                LOCK_DEADLOCK,
+                "Deadlock found when trying to get lock; try restarting transaction",
+            ),
+            (
+                InternalError,
+                LOCK_WAIT_TIMEOUT,
+                "Lock wait timeout exceeded; try restarting transaction",
+            ),
+        ),
+        ids=("deadlock-legacy", "deadlock", "wait-timeout"),
+    )
+    def test_retriable_exceptions(
+        self, celery, redis, session, db, metricsmock, errclass, errno, errmsg
+    ):
+
+        obs = CellObservationFactory.build(radio=Radio.lte)
+        shard = CellShard.shard_model(obs.cellid)
+        cell = CellShardFactory.build(
+            radio=obs.radio,
+            cc=obs.mcc,
+            mnc=obs.mnc,
+            lac=obs.lac,
+            cid=obs.cid,
+            samples=10,
+            created=datetime(2019, 12, 5, tzinfo=UTC),
+        )
+        session.add(cell)
+        session.commit()
+
+        error = errclass(errno, errmsg)
+        wrapped = InterfaceError.instance(
+            statement="SELECT COUNT(*) FROM cell_area",
+            params={},
+            orig=error,
+            dbapi_base_err=MySQLError,
+        )
+        with mock.patch.object(
+            CellUpdater, "add_area_update", side_effect=[wrapped, None]
+        ), mock.patch("ichnaea.data.station.time.sleep") as sleepy:
+            self.queue_and_update(celery, [obs])
+            assert CellUpdater.add_area_update.call_count == 2
+            sleepy.assert_called_once_with(1)
+
+        cells = session.query(shard).all()
+        assert len(cells) == 1
+        self.check_statcounter(redis, StatKey.cell, 1)
+
+        # The existing cell record was updated
+        cell = cells[0]
+        assert cell.samples == 11
+        assert cell.created == datetime(2019, 12, 5, tzinfo=UTC)
+        self.check_statcounter(redis, StatKey.unique_cell, 0)
+
+        # Assert generated metrics are correct
+        assert metricsmock.has_record(
+            "incr", "data.observation.insert", value=1, tags=["type:cell"]
+        )
+        assert metricsmock.has_record(
+            "incr", "data.station.confirm", value=1, tags=["type:cell"]
+        )
+        assert metricsmock.has_record("timing", "task", tags=["task:data.update_cell"])
+        assert metricsmock.has_record(
+            "incr",
+            "data.station.dberror",
+            value=1,
+            tags=["type:cell", "errno:%s" % errno],
+        )
+
+    @pytest.mark.parametrize(
+        "errclass,errno,errmsg",
+        (
+            (
+                OperationalError,
+                DBACCESS_DENIED_ERROR,
+                "Access denied for user 'ichnaea'@'localhost' to database 'location'",
+            ),
```

Comment:
I don't understand what this test is here for. Is this error different than the others? Do we really need to be testing a situation that's a infrastructure/configuration problem?

Reply:
The test is that a non-retryable database error aborts the task. It might show up in a code coverage report. If that's not valuable enough, we can lose the 50 lines.

Reply:
Oh, I see--you used that error because it wasn't retryable. I think that makes sense to test. I thought you were explicitly trying to test an access error which I didn't think was helpful.

Maybe (in a future PR), we could extract the retry logic and make it a function or a decorator or a context manager or something like we have with Antenna and Socorro. Then we can test the retry functionality separate from whatever needs retrying. I don't know if there are other retryable things in the code base.

Anyhow, having this in or out is fine. Whichever you prefer--your feelings on how the code might evolve over time and what kinds of things will help are insightful here.

Reply:
> Maybe (in a future PR), we could extract the retry logic and make it a function or a decorator or a context manager or something like we have with Antenna and Socorro. Then we can test the retry functionality separate from whatever needs retrying. I don't know if there are other retryable things in the code base.

I think only the ``StationUpdater`` has this retry logic, and the Sentry instances of deadlock errors from the 1.5 branch were from other updaters, like the cell area or datamap updaters. So, there may be a chance to make a general way of defining retrying.

There's other things I'd like to try.  Using ``SELECT ... FOR UPDATE`` would move locking earlier in the algorithm, and may eliminate the deadlocks (while increasing lock timeouts). This would make the tradeoffs in chunk size more apparent.

In other projects, I've relied on Celery to rerun the task later on failures. It doesn't work for us because we're pulling data from a Redis queue rather than passing it as arguments. I think this suggests we're using Celery wrong, or maybe Celery is a poor match for this volume of data.

Reply:
> Anyhow, having this in or out is fine. Whichever you prefer--your feelings on how the code might evolve over time and what kinds of things will help are insightful here.

I'm inclined to leave it out for now. If we end up needing something like it, we can resurrect this test code.

Reply:
> In other projects, I've relied on Celery to rerun the task later on failures. It doesn't work for us because we're pulling data from a Redis queue rather than passing it as arguments. I think this suggests we're using Celery wrong, or maybe Celery is a poor match for this volume of data.

Celery was never really a good fit for this project, but it was simple to get going and ops had experience with it. The data pipeline needs to be able to split/multiplex and change batch sizes of the incoming data (e.g. commit in ~1mb blocks to AWS S3 as each put request costs money and has size limitations, but the same data should be processed faster for the SQL backend). None of those things were available in Celery at the time, so all the custom Redis logic was written. The code was probably influenced by me working on a Kafka based project before. If I had been able to continue work on the project, I’d have probably converted it to use AWS Kinesis for the data pipeline part and AWS Aurora as a MySQL drop-in. In a later step, I would have replaced the database side with a key/value store for the main data and some smaller db for the statistics side.